File: ui/src/lib/api/policies.ts
import { api } from '../fetcher';
export interface PolicyRule { id:number; version:number; name:string; group:string; action:string; enabled:boolean; runtime_ready:boolean }
export interface PolicyDocument { name:string; description:string; host:string; path_prefix:string; mode:'mixed'|'block'|'detect'; priority:number; rule_ids:number[]; rules:PolicyRule[] }
export interface SavedPolicy { id:number; version:number; published_version:number|null; status:'Draft'|'Published'|'Disabled'; document:PolicyDocument; created_at:string; updated_at:string; actor?:string; operation?:string }
export interface PolicyRelease { release_id:number; digest:string; payload:unknown; membership:{id:number;version:number;name:string}[]; preview:boolean }
export interface ClusterPolicies { release_id:number; nodes:{id:string;release_id:number;phase:string;message:string;updated_at:string}[] }
export type PolicyDraft = Omit<PolicyDocument,'rules'> & { expected_version:number };
export const policiesApi = {
 list:()=>api.get<SavedPolicy[]>('/api/v1/policies'),
 detail:(id:number,history=false)=>api.get<SavedPolicy[]>(`/api/v1/policies/${id}`,{history}),
 catalog:()=>api.get<PolicyRule[]>('/api/v1/policies/catalog'),
 cluster:()=>api.get<ClusterPolicies>('/api/v1/policies/cluster'),
 save:(id:number|null,body:PolicyDraft,key:string)=>id?api.put<{id:number;version:number}>(`/api/v1/policies/${id}`,body,{idempotencyKey:key}):api.post<{id:number;version:number}>('/api/v1/policies',body,{idempotencyKey:key}),
 restore:(id:number,version:number,restore:number,key:string)=>api.put(`/api/v1/policies/${id}`,{expected_version:version,restore_version:restore},{idempotencyKey:key}),
 publish:(id:number,version:number,head:number,disable:boolean,key:string,preview=false)=>api.post<PolicyRelease>(`/api/v1/policies/${id}/publish${preview?'?preview=true':''}`,{expected_version:version,expected_release:head,disable},{idempotencyKey:key}),
};

File: control-plane/internal/domain/entity/policies.go
package entity

import "encoding/json"

// Draft mutations capture rule revisions, never references to mutable live rules.
type SavePolicyCommand struct {
	ID              int64   `json:"-"`
	ExpectedVersion int64   `json:"expected_version"`
	RestoreVersion  int64   `json:"restore_version,omitempty"`
	Name            string  `json:"name"`
	Description     string  `json:"description"`
	Host            string  `json:"host"`
	PathPrefix      string  `json:"path_prefix"`
	Mode            string  `json:"mode"`
	Priority        int     `json:"priority"`
	RuleIDs         []int64 `json:"rule_ids"`
	Actor           string  `json:"-"`
	RequestKey      string  `json:"-"`
}
type SavePolicyResult struct {
	ID      int64 `json:"id"`
	Version int64 `json:"version"`
}
type ReadPoliciesQuery struct {
	ID      int64
	History bool
}
type ReadPoliciesItem struct {
	Status           string          `json:"status"`
	ID               int64           `json:"id"`
	Version          int64           `json:"version"`
	PublishedVersion *int64          `json:"published_version"`
	Document         json.RawMessage `json:"document"`
	CreatedAt        string          `json:"created_at"`
	UpdatedAt        string          `json:"updated_at"`
	Actor            string          `json:"actor,omitempty"`
	Operation        string          `json:"operation,omitempty"`
}
type PolicyCatalogItem struct {
	ID   int64  `json:"id"`
	Name string `json:"name"`
}
type PolicyCatalogRule struct {
	ID           int64  `json:"id"`
	Version      int64  `json:"version"`
	Name         string `json:"name"`
	Group        string `json:"group"`
	Action       string `json:"action"`
	Enabled      bool   `json:"enabled"`
	RuntimeReady bool   `json:"runtime_ready"`
}
type PublishPolicyCommand struct {
	ID              int64  `json:"-"`
	ExpectedVersion int64  `json:"expected_version"`
	ExpectedRelease int64  `json:"expected_release"`
	Disable         bool   `json:"disable"`
	Preview         bool   `json:"-"`
	Actor           string `json:"-"`
	RequestKey      string `json:"-"`
}
type PublishPolicyResult struct {
	ReleaseID  int64           `json:"release_id"`
	Digest     string          `json:"digest"`
	Payload    json.RawMessage `json:"payload"`
	Membership json.RawMessage `json:"membership"`
	Preview    bool            `json:"preview"`
}
type PolicyClusterStatus struct {
	ReleaseID int64               `json:"release_id"`
	Nodes     []PolicyClusterNode `json:"nodes"`
}
type PolicyClusterNode struct {
	ID        string `json:"id"`
	ReleaseID int64  `json:"release_id"`
	Phase     string `json:"phase"`
	Message   string `json:"message"`
	UpdatedAt string `json:"updated_at"`
}
type PolicySyncQuery struct{ NodeID string }
type PolicySyncResult struct {
	ReleaseID int64           `json:"release_id"`
	Digest    string          `json:"digest"`
	Payload   json.RawMessage `json:"payload"`
}
type PolicyReportCommand struct {
	NodeID    string `json:"-"`
	ReleaseID int64  `json:"release_id"`
	Phase     string `json:"phase"`
	Message   string `json:"message"`
}
